Python/pywarpx/diagnostics/data/BaseReader.py

```diff
@@ -0,0 +1,173 @@
+import collections
+import os
+import re
+
+import numpy as np
+import pandas as pd
+
+
+class DataReader():
+    """
+    Base class that all data readers should inherit from.
+    Works only in Cartesian geometry.
+    """
+
+    def __init__(self, run_directory, file_prefix=None):
+        """
+        Parameters
+        ----------
+        run_directory: string
+            Path to the run directory of WarpX.
+        """
+        if run_directory is None:
+            raise ValueError('The run_directory parameter can not be None!')
+
+        self.run_directory = run_directory
+        self.data_file_prefix = None
+        self.data_file_suffix = None
+
+        self.has_species = None
+
+    def get_data_path(self, subdir='reducedfiles'):
+        """
+        Returns the path to the data file.
+        Parameters
+        ----------
+        subdir: string
+            Name of reducedfiles directory, default is 'reducedfiles'.
+        Returns
+        -------
+        A string with a path to the current diagnostic (assumed unique).
+        """
+
+        sim_output_dir = os.path.join(self.run_directory, 'diags/'+subdir+'/')
+        if not os.path.isdir(sim_output_dir):
+            raise IOError('The simulation directory does not exist inside '
+                          'path:\n  {}\n'
+                          'Did you set the proper path to the run directory?\n'
+                          'Did the simulation already run?'
+                          .format(self.run_directory))
+
+        data_file_path = os.path.join(
+            sim_output_dir,
+            self.data_file_prefix +
+            self.data_file_suffix
+        )
+        if not os.path.isfile(data_file_path):
+            raise IOError('The file {} does not exist.\n'
+                          'Did the simulation already run?'
+                          .format(data_file_path))
+
+        return data_file_path
+
+    def get_ncols(self):
+        ''' Returns the number of columns of the reduced diagnostic.'''
+        data_file_path = self.get_data_path()
+        data = np.loadtxt(data_file_path)
```

Comment:
This may not matter much in general, but if the data files become large, it will be more efficient to read in the data once rather than the way here where the data is read each time a query is made.
Also, are there any cases where ndim == 1, since the files always include the step and time?

Reply:
Makes sense. What about if it only reads the first column? 
`ndim = 1` was meant for the case where only 1 timestep is dumped. 
Anyway the following should solve all these issues:
```py
header=pd.read_csv(data_file_path, header=0, sep=" ', nrows=0)
ncols = len(header.columns)
```
Do you agree?

Reply:
Yes, that would be good, using pandas here.
Note that another advantage of reading in the data once is that the loadtxt or read_csv command appear in only one place, reducing code duplication.

Reply:
I understand you point about loading the data just once, but I'm not sure I understand how to implement it given the current design.  

Right now the idea is that the user can directly extract the desired data without having to load the whole file.
Anytime the user calls a `get_data` kind of function, the whole file will be loaded and the desired data will be selected. 
The user can also ask for the available steps and times. 
If we keep this logic, I'm not sure how to have just one occurrence of loadtxt or read_csv command. 

Another approach would be to make the user load the whole file on their own at first, and after that use these tools to extract the desired data. 
To give an idea, something like this:
```
all_data = diag.load(filename) # here the whole file is loaded
steps  = all_data.steps()
desired_data = all_data.get_data()
```
Do you think the latter is a better rationale?

Reply:
I think it would be fine to have methods that query the file without reading in all of the data, for example reading only the headers or reading in one line of data to get the number of columns. For anything else though that would process the whole file, I think it would be simpler and faster to just read in the data once and query the result as needed. However, it is unlikely that the files will be truly large so just reading in the data right away would likely not be an issue. (My timings show that loadtxt and pandas can read in 1,000,000 lines of data in under a second.)